### ISSUE TITLE
Uncaught error in rebar_core - undef function

### ISSUE THREAD

**Issue by pulitta** [blt/port_compiler#43]
Problem with the latest version of pc (v1.7.1).
```
===> Uncaught error in rebar_core. Run with DEBUG=1 to see stacktrace or consult rebar3.crashdump
===> Uncaught error: undef
===> Stack trace to the error location:
[{rebar_string,chr,["priv/scrypt_port/*.c",36],[]},
 {pc_port_specs,expand_env,2,
                [{file,".../_build/default/plugins/pc/src/pc_port_specs.erl"},
                 {line,182}]},
 {pc_port_specs,'-get_port_spec/3-fun-0-',3,
                [{file,".../_build/default/plugins/pc/src/pc_port_specs.erl"},
                 {line,154}]},
 {pc_port_specs,'-get_port_spec/3-lists^flatmap/1-0-',2,
                [{file,".../_build/default/plugins/pc/src/pc_port_specs.erl"},
                 {line,152}]},
 {pc_port_specs,get_port_spec,3,
                [{file,".../_build/default/plugins/pc/src/pc_port_specs.erl"},
                 {line,152}]},
 {pc_port_specs,'-construct/1-lc$^0/1-1-',2,
                [{file,".../_build/default/plugins/pc/src/pc_port_specs.erl"},
                 {line,67}]},
 {pc_port_specs,construct,1,
                [{file,".../_build/default/plugins/pc/src/pc_port_specs.erl"},
                 {line,67}]},
 {pc_prv_compile,'-do/1-lc$^0/1-0-',2,
                 [{file,".../_build/default/plugins/pc/src/pc_prv_compile.erl"},
                  {line,52}]}]
```

**Comment by tsloughter** [comment]
What version of rebar3 are you using?

**Comment by pulitta** [comment]
rebar 3.3.5 on Erlang/OTP 19 Erts 8.1

**Comment by tsloughter** [comment]
Please try 3.5 or above. There were changes necessary for future OTP versions that change the string module and I think this issue stems from those changes.

**Comment by jeffjewiss** [comment]
Hi, I think I'm running into a similar issue.

I'm trying to run tests with CirclCI and pc v1.6 worked fine, but with pc v1.7.1 I get:
```
(Mix) Could not compile dependency :erlexec, "/home/circleci/.mix/rebar3 bare compile --paths "/home/circleci/repo/_build/test/lib/*/ebin"" command failed. You can recompile this dependency with "mix deps.compile erlexec", update it with "mix deps.update erlexec" or clean it with "mix deps.clean erlexec"
```

I'm using: `rebar 3.5.0 on Erlang/OTP 20 Erts 9.3`

**Comment by isaiahdw** [comment]
Anyone have any luck figuring out this issue? It seems like it happens with both rebar 3.5.0 and rebar 3.4.1 (both tested with Erlang/OTP 20).

**Comment by rkallos** [comment]
Same issue here. When switching to rebar 3.5.0, I get a different error:
```
===> sh info:
	cwd: "/home/richard/work/envo/envo"
	cmd: g++ -c  -DHAVE_SETRESUID -DHAVE_CAP -DHAVE_PTRACE -std=c++98 -O0 -g -Wall -fPIC -MMD  -I"/usr/local/stow/OTP-20.2.2/lib/erlang/lib/erl_interface-3.10.1/include" -I"/usr/local/stow/OTP-20.2.2/lib/erlang/erts-9.2/include"   /home/richard/work/envo/envo/_build/default/lib/erlexec/c_src/*.cpp -o /home/richard/work/envo/envo/_build/default/lib/erlexec/c_src/*.o

[snip]
===> Compiling /home/richard/work/envo/envo/_build/default/lib/erlexec/c_src/*.cpp
===> g++: fatal error: cannot specify -o with -c, -S or -E with multiple files
compilation terminated.
```

It appears that using pc 1.7.1 now calls g++ for compiling erlexec with `-c`, which fails when using multiple input files (as is the case with erlexec). This didn't appear to happen when using pc 1.6.0. Looking through the changes between pc 1.7.1 and 1.6.0, I can't figure out what change introduces this behavior. My guess: pc 1.6.0 did not respect `DRV_CXX_TEMPLATE`, which has had `-c` for a long time, and now it does.

**Comment by pulitta** [comment]
Switched to rebar 3.5.0 and got different error:
```
===> Compiling .../_build/default/lib/scrypt/priv/scrypt_port/*.c
===> clang: error: cannot specify -o when generating multiple output files
```

**Comment by tsloughter** [comment]
ugh, sorry about this, seems I shouldn't have published this version. And it doesn't seem like it is just a minor version bump if it breaks this much.

@tuncer what is going on?

**Comment by tsloughter** [comment]
Note you can try replacing the pc plugin with a specific version, `{plugins, [{pc, "1.6.0"}]}.`

**Comment by tsloughter** [comment]
If it is a dependency you can use overrides:

```
{overrides, [{override, <DEP>, [{plugins, [{pc, "1.6.0"}]}}]}.
```

**Comment by dlameri** [comment]
I can't do that, since i'm importing ejabberd, that imports fast_tls (i think), that then imports pc, on a build script. There is no way to override it.

**Comment by tsloughter** [comment]
@dlameri not sure what you mean. If you are using rebar3 then you can add an override in your rebar config.

**Comment by tsloughter** [comment]
I just retired 1.7.0 and 1.7.1 and published the old version again as 1.8.0.

`rebar3 update` to get the latest registry and then `pc` will automatically get 1.8.0. If your project has already fetched pc then you'll need to do `rebar3 plugins upgrade pc`. Or blow away `_build`.

**Comment by isaiahdw** [comment]
@tsloughter, it works great now. Thanks for the quick fix!

**Comment by jeffjewiss** [comment]
Thanks for the quick fix on this!

**Comment by ghost** [comment]
@tsloughter, the illegal call to `rebar_string:chr/2` was introduced in https://github.com/blt/port_compiler/pull/38.

**Comment by ghost** [comment]
@rkallos, the plugin wasn't aware of the need to use CXX when building C++ sources, and this worked because users added extra flags, basically imitating what the CXX frontend/compile-driver does internally. This worked, but isn't reliable and definitely the wrong thing to do. That behavior has existed since the inception of the port_compiler plugin in 2009. As you noticed, current port_compiler can detect C++ sources and call CXX for those, which is the reliable and right thing to do.

**Comment by ghost** [comment]
See #45.

**Comment by tuncer** [comment]
Merged #45 and tagged `v1.8.2`. @tsloughter, can you publish on hex?

**Comment by ghost** [comment]
Merged #45 and tagged `v1.8.3`. @tsloughter, can you publish on hex?

**Comment by ghost** [comment]
Fixed #26 as well and tagged `v1.9.1`. Let's publish that on hex.
